python/ql/src/Security/CWE-327/TlsLibraryModel.qll

```diff
@@ -0,0 +1,134 @@
+import python
+import semmle.python.ApiGraphs
+import Ssl
+import PyOpenSSL
+
+/**
+ * A specific protocol version.
+ * We use this to identify a protocol.
+ */
+class ProtocolVersion extends string {
+  ProtocolVersion() { this in ["SSLv2", "SSLv3", "TLSv1", "TLSv1_1", "TLSv1_2", "TLSv1_3"] }
+
+  predicate lessThan(ProtocolVersion version) {
+    this = "SSLv2" and version = "SSLv3"
+    or
+    this = "TLSv1" and version = ["TLSv1_1", "TLSv1_2", "TLSv1_3"]
+    or
+    this = ["TLSv1", "TLSv1_1"] and version = ["TLSv1_2", "TLSv1_3"]
+    or
+    this = ["TLSv1", "TLSv1_1", "TLSv1_2"] and version = "TLSv1_3"
+  }
+
+  predicate isInsecure() { this in ["SSLv2", "SSLv3", "TLSv1", "TLSv1_1"] }
+}
+
+/** An unspecific protocol version */
+class ProtocolFamily extends string {
+  ProtocolFamily() { this in ["SSLv23", "TLS"] }
+}
+
+/** The creation of a context. */
+abstract class ContextCreation extends DataFlow::CfgNode {
+  /** Gets the requested protocol if any. */
+  abstract DataFlow::CfgNode getProtocol();
+}
+
+/** The creation of a connection from a context. */
+abstract class ConnectionCreation extends DataFlow::CfgNode {
+  /** Gets the context used to create the connection. */
+  abstract DataFlow::CfgNode getContext();
+}
+
+/** A context is being restricted on which protocols it can accepts. */
+abstract class ProtocolRestriction extends DataFlow::CfgNode {
+  /** Gets the context being restricted. */
+  abstract DataFlow::CfgNode getContext();
+
+  /** Gets the protocol version being disallowed. */
+  abstract ProtocolVersion getRestriction();
+}
+
+/** A context is being relaxed on which protocols it can accepts. */
+abstract class ProtocolUnrestriction extends DataFlow::CfgNode {
+  /** Gets the context being relaxed. */
+  abstract DataFlow::CfgNode getContext();
+
+  /** Gets the protocol version being allowed. */
+  abstract ProtocolVersion getUnrestriction();
+}
+
+abstract class UnspecificContextCreation extends ContextCreation, ProtocolUnrestriction {
+  TlsLibrary library;
+  ProtocolFamily family;
+
+  UnspecificContextCreation() { this.getProtocol() = library.unspecific_version(family) }
+
+  override DataFlow::CfgNode getContext() { result = this }
+
+  override ProtocolVersion getUnrestriction() {
+    family = "TLS" and
+    result in ["TLSv1", "TLSv1_1", "TLSv1_2", "TLSv1_3"]
+    or
+    // This can negotiate a TLS 1.3 connection (!)
+    // see https://docs.python.org/3/library/ssl.html#ssl-contexts
+    family = "SSLv23" and
+    result in ["SSLv2", "SSLv3", "TLSv1", "TLSv1_1", "TLSv1_2", "TLSv1_3"]
```

Comment:
This seems wrong to me, since `PROTOCOL_TLS` and `PROTOCOL_SSLv23` are just aliases. See https://docs.python.org/3.9/library/ssl.html#ssl.PROTOCOL_SSLv23

Reply:
It seems they are not exactly aliases everywhere. I added some references in comments.

Reply:
I raised this in a different comment, so won't re-open this one